Fix and improvment of Swiss Hydrological Data component

Bouni opened the pull request:
## Description:

Fix for the not working Swiss Hydrological Data component.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6535

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: swiss_hydrological_data
    station: 2143
    name: Rhein Rekingen
    measurements:
      - temperature
      - level
      - discharge
      - min_temperature
      - min_level
      - min_discharge
      - max_temperature
      - max_level
      - max_discharge
      - mean_temperature
      - mean_level
      - mean_discharge
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Bouni committed:
Fix and improvment of Swiss Hydrological Data component

Bouni committed:
Fix and improvment of Swiss Hydrological Data component

homeassistant commented:
Hi @Bouni, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#17166).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented:
The issue with the proposed solution is that every instance is going to download around 800 kB of data per hour (approx. 30-40 kB per file) when all sensors are activate but only one value per file is of interest. Thus we are wasting a lot of resources.

When I wrote the integration the FOEN agreed on a automatic approach to get the XML file with the data  from ftp.hydrodata.ch. Now the access requires credentials which made our integration non-functional.

I see three way to fix it:

1. Got back to the scrape approach which was used in the first place and I provided to some users.
2. Get the XML file and publish the content by ourself as RESTFul API
3. Team-up with the guys from https://aare.guru. Last time we spoke they were interested in sharing the data because they have access to the XML file.

Bouni commented:
@fabaff 
Totally agree with you! To be honest I wasn't aware of the size of the CSV files!
I'll take a look into your suggestions and update the pull request as soon as I've found a viable solution!

Bouni commented:
@fabaff 
I've got in touch with FEON to ask them about the HTTP Interface the seem to [offer](https://www.hydrodata.ch/).
- They no longer provide the HTTP nor the anonymous interface
- They switched to HTTPS and a per user account
- They don't alow to use their server as backend and limit the request rate to once every 10 minutes.
- They told me that they will change their XML format in 2020 but will be backwards compatible.
- They normally give no access to private persons but just to specialist agencies.

Anyway, the signaled that if we could have our own backend (as you've suggested) we could get an account for that.

What exactly did you mean by "publish the content by ourself as RESTFul API" where would that service be hosted? Its obviously not a big deal writing the interpreter and REST API, but who is going to host it? Or are there already other components relying on such an HA provided API which could be implemented for this component as well?

fabaff commented:
If you give me access to the API then I can create a simple client which we then can use for this sensor.

Bouni commented:
@fabaff I still wait for the credentials, at the moment the API just respondes data from a single static file they gave me for reference. I'll ping you as soon as the API is live!

Bouni commented:
@fabaff I think I'm done with my API but I still wait for the login data from FEON.
You can have a look at the API (with a static xml at the moment): https://swisshydroapi.bouni.de/

> If you give me access to the API then I can create a simple client which we then can use for this sensor.
What exactly do you mean by "simple client" ?

Bouni commented:
Just got an E-Mail, my account at FEON will be ready by 5th of November.
I've no idea what takes that long toset up another account, but thats how it works I assume :hear_no_evil:

fabaff commented:
> What exactly do you mean by "simple client" ?

A module that is performing the requests and process the received data.

Bouni commented:
I've already done that with this https://pypi.org/project/swisshydrodata/ 
Yesterday in the evening I got my access data from FEON for their data service and since then my API is up and running.
![grafik](https://user-images.githubusercontent.com/948965/48048493-fb792080-e19b-11e8-8d57-92d7569d72f5.png)

I would like to test the component for a few days to see if there are any problems and then update my pull request. Is there anything you want me to do / to test before I update the PR?

fabaff commented:
> I've already done that with this https://pypi.org/project/swisshydrodata/

Nice

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -12,137 +12,126 @@
-            attributes[ATTR_LOCATION] = self.data.measurings['location']
-            attributes[ATTR_UPDATE] = self.data.measurings['update_time']
-            attributes[ATTR_ATTRIBUTION] = CONF_ATTRIBUTION
-            return attributes
```
This is a breaking change which needs an entry in the release notes.

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -12,137 +12,126 @@
-            attributes[ATTR_ATTRIBUTION] = CONF_ATTRIBUTION
-            return attributes
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION
```
How about adding the station details? Location, etc.

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -151,38 +140,29 @@ class HydrologicalData:
+        data["max_discharge"] = shd.get_max_discharge()
+        data["mean_temperature"] = shd.get_mean_temperature()
+        data["mean_level"] = shd.get_mean_level()
+        data["mean_discharge"] = shd.get_mean_discharge()
```
I'm not so sure that this is the right approach resource-wise. There will be a dozen requests for every update to get the data instead of one that is sharing the data.

You want to change the sensor style (single sensors instead of attributes), thus you should only get the data, e. g., for `mean_temperature` that's needed and not everything that is available. BTW, there is a chance (https://github.com/home-assistant/architecture/issues/100) that it needs to be updated/changed again soon.

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -12,137 +12,126 @@
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Required(CONF_STATION): vol.Coerce(int),
     vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_MEASUREMENTS): vol.Any([
```
That's a breaking change. Please check the documentation for the validation to check that it's a list.

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -12,137 +12,126 @@
-        _LOGGER.error("The URL is not accessible")
+    data = HydrologicalData(station)
+
+    response = requests.get(
```
Should be tested with `swisshydrodata` as `shd` should be created here anyway and passed to `HydrologicalData`.

fabaff commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -12,137 +12,126 @@
+    )
+    if response.status_code != 200:
+        _LOGGER.error("The given station does not exist: %s", station)
         return False
```
Can be just `return`.

Bouni committed:
changed component to get data from a REST API rather than from crawling the website

Bouni commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -8,141 +8,158 @@
         """Return the unit of measurement of this entity, if any."""
         if self._state is not STATE_UNKNOWN:
-            return self._unit_of_measurement
+            return self.hydro_data.data["parameters"][self._measurement]["unit"]
```
@fabaff Whats the pythonic way for breaking lines like this so that I meet the hounds requirements?

fabaff replied:
There seems no other way than between the brackets here.

Bouni replied:
I tried several ways and I always get lint errors, so I used this "trick":
https://github.com/home-assistant/home-assistant/blob/0dd43bb424ef846f18cc33d8dd28873fcf38f452/homeassistant/components/sensor/swiss_hydrological_data.py#L138-L142

Bouni committed:
fixed several issues and lint errors

Bouni commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -29,67 +28,44 @@
+    vol.Required(CONF_MONITORED_CONDITIONS): vol.Schema({
+        "temperature": vol.All(cv.ensure_list, [vol.In(CONDITIONS)]),
+        "level": vol.All(cv.ensure_list, [vol.In(CONDITIONS)]),
+        "discharge": vol.All(cv.ensure_list, [vol.In(CONDITIONS)])
```
@fabaff Is this what you meant by ensuring that the conditions are a list?

Bouni commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -111,33 +88,35 @@ def __init__(self, hydro_data, measurement, value):
-            self._measurement,
+            self._station,
+            self._condition,
             self._value)
```
@fabaff What do you think about the way I name the sensors? Is that a good way?

Bouni replied:
The "problem" is that the waterbody name is something I get from the API and if its down for some reason I cannot know the waterbody name when setting up the entities. Thats why I went for the station ID. The user still can set a friendly_name, right!?

Bouni commented on homeassistant/components/sensor/swiss_hydrological_data.py:
```diff
@@ -111,33 +88,35 @@ def __init__(self, hydro_data, measurement, value):
+            "ATTR_UPDATE":
+            self.hydro_data.data["parameters"][self._condition]["datetime"],
             "ATTR_ATTRIBUTION": CONF_ATTRIBUTION
         }
```
I'm not sure whats the prefered way to set attributes and I'm not very happy with this, especially how they show up in HA (Lovelace UI)
![grafik](https://user-images.githubusercontent.com/948965/48117963-12d30f00-e26b-11e8-8a22-668216cdd256.png)

fabaff replied:
You removed the definition of all constants (line 33-42) and made the key strings in the dict.

Bouni replied:
Ok, got that, I'll fix it today if I find some spare time.

Bouni commented:
:hankey: I think I've messed someing up

fabaff committed:
Minor changes

- Simplify the sensor configuration (expose value as attributes rather than sensor)
- Make the setup fail if station is not available
- Add unique ID
- Prepare for config flow

fabaff reviewed (approved):
Looks good to me :bird:

swissuser31415 commented:
- platform: swiss_hydrological_data
    station: 6600
    monitored_conditions:
      - temperature


![image](https://user-images.githubusercontent.com/84203216/118770814-0db29a00-b882-11eb-848b-66c66dcdb2d8.png)


Any ideas? The 6600 station does exist according to the website and delivers the temperature data.

frenck commented:
@swissuser31415 this is a pull request from 2018, please don't bump old/closed/merged/handled PRs. 

If you need support, please see our website on where to go:

<https://www.home-assistant.io/help>

Thanks! 👍
